R/parse-body.R

```diff
@@ -76,54 +94,92 @@ parser_picker <- function(content_type, first_byte, filename = NULL) {
+#' corresponding parsing function.
+#' Functions signature in the list should include `value`, `...` and
+#' possibly `content_type`, `filename`. Other parameters may be provided
+#' if you want to use the headers from [webutils::parse_multipart()].
+#' Parser function structure is something like below.
 #' ```r
-#' parser <- function(value, content_type = "ct", filename, ...) {
+#' parser <- () {
+#'  f <- function(value, ...) {
 #'   # do something with raw value
+#'  }
+#'  make_parsers(f, fixed = "ct")
 #' }
 #' ```
 #'
-#' It should return a named list if you want values to map to
-#' plumber endpoint function args.
-#'
 #' @examples
-#' parser_dcf <- function(value, content_type = "text/x-dcf", ...) {
-#'   charset <- getCharacterSet(content_type)
-#'   value <- rawToChar(value)
-#'   Encoding(value) <- charset
-#'   read.dcf(value)
+#' # Content-type header is mostly used to look up charset and adjust encoding
+#' parser_dcf <- function() {
+#'   f <- function(value, content_type = "text/x-dcf", ...) {
+#'     charset <- getCharacterSet(content_type)
+#'     value <- rawToChar(value)
+#'     Encoding(value) <- charset
+#'     read.dcf(value)
+#'   }
+#'   return(make_parsers(f, fixed = "text/x-dcf"))
 #' }
+#' add_parser("dcf", parser_dcf)
 #' @export
-add_parser <- function(content_type, parser, verbose = TRUE) {
+add_parser <- function(alias, parser, verbose = TRUE) {
 
-  if (!is.null(.globals$parsers[[content_type]])) {
+  if (!is.null(.globals$parsers[[alias]])) {
     if (isTRUE(verbose)) {
-      warning("Overwriting parser: ", content_type)
+      warning("Overwriting parser: ", alias)
     }
   }
 
   stopifnot(is.function(parser))
 
-  .globals$parsers[[content_type]] <- parser
+  .globals$parsers[[alias]] <- parser
 
-  invisible(.globals$parsers)
+  invisible(list_parsers())
 }
 
+#' @export
+#' @describeIn add_parser List currently registered parsers
+list_parsers <- function() {
+  .globals$parsers
+}
+
+#' @export
+#' @describeIn add_parser Make named list. Mapping content-type with parser.
+#' @param parser_function A single functions to map to one or more Content-Type.
+#' @param fixed A character vector of fixed string to be matched against a request Content-Type.
+#' @param regex A character vector of [regex] string to be matched against a request Content-Type.
+#' @param shortname A character value to reference a parser by a shortname.
+#' Content-Type.
+make_parsers <- function(parser_function, fixed = NULL, regex = NULL, shortname = NULL) {
+  if (any(shortname %in% c("fixed", "regex"))) {
+    stop("Shortnames `fixed` and `regex` are reserved for internal use.")
+  }
+  m <- function(n) {
+    structure(
+      replicate(
+        length(n),
+        parser_function),
+      names = n
+    )
+  }
+  parsers <- m(shortname)
+  if (length(fixed) > 0) {
+    parsers <- c(parsers, list("fixed" = m(fixed)))
+  }
+  if (length(regex) > 0) {
+    parsers <- c(parsers, list("regex" = m(regex)))
+  }
```

Comment:
Small cosmetic things...

```suggestion
  create_list <- function(names) {
    stats::setNames(
      replicate(
        length(names),
        parser_function),
      names
    )
  }
  parsers <- create_list(shortname)
  if (length(fixed) > 0) {
    parsers$fixed <- create_list(fixed)
  }
  if (length(regex) > 0) {
    parsers$regex <- create_list(regex)
  }
```